src/Nethermind/Nethermind.Evm/Instructions/EvmInstructions.CodeCopy.cs

```diff
@@ -0,0 +1,297 @@
+    }
+
+    /// <summary>
+    /// Retrieves the size of the external code of an account.
+    /// Pops an account address from the stack, validates access, and pushes the code size onto the stack.
+    /// Additionally, applies peephole optimizations for common contract checks.
+    /// </summary>
+    /// <typeparam name="TTracingInstructions">
+    /// A struct implementing <see cref="IFlag"/> indicating if instruction tracing is active.
+    /// </typeparam>
+    /// <param name="vm">The virtual machine instance.</param>
+    /// <param name="stack">The EVM stack from which the account address is popped and where the code size is pushed.</param>
+    /// <param name="gasAvailable">Reference to the available gas; reduced by external code cost.</param>
+    /// <param name="programCounter">Reference to the program counter, which may be adjusted during optimization.</param>
+    /// <returns>
+    /// <see cref="EvmExceptionType.None"/> on success, or an appropriate error code if an error occurs.
+    /// </returns>
+    [SkipLocalsInit]
+    public static EvmExceptionType InstructionExtCodeSize<TTracingInstructions>(
+        VirtualMachine vm,
+        ref EvmStack stack,
+        ref long gasAvailable,
+        ref int programCounter)
+        where TTracingInstructions : struct, IFlag
+    {
+        IReleaseSpec spec = vm.Spec;
+        // Deduct the gas cost for external code access.
+        gasAvailable -= spec.GetExtCodeCost();
+
+        // Pop the account address from the stack.
+        Address address = stack.PopAddress();
+        if (address is null) goto StackUnderflow;
+
+        // Charge gas for accessing the account's state.
+        if (!ChargeAccountAccessGas(ref gasAvailable, vm, address))
+            goto OutOfGas;
+
+        // Attempt a peephole optimization when tracing is not active and code is available.
+        ReadOnlySpan<byte> codeSection = vm.EvmState.Env.CodeInfo.MachineCode.Span;
+        if (!TTracingInstructions.IsActive && programCounter < codeSection.Length)
+        {
+            bool optimizeAccess = false;
+            // Peek at the next instruction to detect patterns.
+            Instruction nextInstruction = (Instruction)codeSection[programCounter];
+            // If the next instruction is ISZERO, optimize for a simple contract check.
+            if (nextInstruction == Instruction.ISZERO)
+            {
+                optimizeAccess = true;
+            }
+            // If the next instruction is GT or EQ and the top stack element is zero,
+            // then this pattern likely corresponds to a contract existence check.
+            else if ((nextInstruction == Instruction.GT || nextInstruction == Instruction.EQ) &&
+                     stack.PeekUInt256IsZero())
+            {
+                optimizeAccess = true;
+                // Remove the zero from the stack since we will have consumed it.
+                if (!stack.PopLimbo()) goto StackUnderflow;
```

Comment:
why goto better than direct return?

Reply:
Direct return adds a return at that place; so it would be the predicted path for cold branch prediction; whereas this is more like throwing an exception and should rarely happen, so we want it to be a jump forward

<img width="530" alt="image" src="https://github.com/user-attachments/assets/5a0a5cc2-2106-4a7a-870c-5fc8dec015ac" />